vnext/ReactUWP/Modules/NativeUIManager.h

```diff
@@ -97,6 +104,8 @@ class NativeUIManager : public facebook::react::INativeUIManager {
   std::vector<winrt::Windows::UI::Xaml::FrameworkElement::SizeChanged_revoker>
       m_sizeChangedVector;
   std::vector<std::function<void()>> m_batchCompletedCallbacks;
+
+  std::map<int64_t, std::weak_ptr<IXamlReactControl>> m_tagsToXamlReactControl;
```

Comment:
>m_tagsToXamlReactControl [](start = 54, length = 24)

The UIManager already has a registry of all tags and you can ask it for all of the root tags and also get the shadow node from a tag which can then provide the XamlView. So, you shouldn't need to keep this map here. You can access this info in this class through m_host, using GetAllRootTags() and FindShadowNodeForTag().

Reply:
Currently we can't deduce ReactControl from ShadowNode directly. It's possible to save IReactRootView in RootShadownNode, but the lifecycle maybe a problem since current interface is 'facebook::react::IReactRootView *rootView' other than std::weak_ptr or shared_ptr.
Also I'm thinking if it's possible to have two NativeUIManagers in the future. Since it's very easy to do it in NativeUIManager, I choose the simple way.

---
In reply to: [299686853](https://github.com/microsoft/react-native-windows/pull/2706#discussion_r299686853) [](ancestors = 299686853)

Reply:
Oof, that's unfortunate. But it's also a reminder that this area should really be refactored at some point. The Native/UI Manager split was really built on top of what was done for Win32, but the XAML's needs didn't match that as well as I'd like.

---
In reply to: [299694541](https://github.com/microsoft/react-native-windows/pull/2706#discussion_r299694541) [](ancestors = 299694541,299686853)

Reply:
What's the benefits of putting it into RootShadowNode? Will we have two native ui managers: one for UWP and one for win32 and they are in the same application?
We we have react::uwp::RootShadowNode and react::other::RootShadowNode? If yes, keeping the mapping in nativemanager may be still a good solution. The only thing  we need to do is to extract this part to a base class and then all other nativemanagers as a subclass of it.
Put it to RootShadowNode would be a headache because we may need more information to cast facebook::react::IReactRootView to react::uwp::IReactRootView, or to react::win32::IReactRootView.
I may create a task to track your requirement after you provide more details.

---
In reply to: [299725037](https://github.com/microsoft/react-native-windows/pull/2706#discussion_r299725037) [](ancestors = 299725037,299694541,299686853)